### PR TITLE
#37 Add Logic to start VsCode as Admin with GitExtensions.SolutionRunner

### LINKED ISSUE #37: Feature Request: Open VsCode work as Run as Administrator too
It would be great if this plugin also could run *.code-workspace files using VSCode in admin mode too.

*I've created PR: https://github.com/maraf/GitExtensions.SolutionRunner/pull/38

### PR DESCRIPTION
**Pull request by AngieKadnar**
Fixes https://github.com/maraf/GitExtensions.SolutionRunner/issues/37

### CODE REVIEW

**Comment by maraf** [comment]
Thank you for contribution! I'll go over the PR over weekend and publish a new version

**Comment by maraf** [comment]
@AngieKadnar I'm getting an error when trying to run VS as admin, the same works in current version. Can you please take a look?

image

**Comment by AngieKadnar** [comment]
@maraf If you activate the RunAsAdmin you need to specify an executable path! Without it throw this error :).
For Example: "C:\Program Files\Microsoft Visual Studio\2022\Enterprise\Common7\IDE\devenv.exe"

image

**Comment by maraf** [comment]
Hmm, why does it seem to work in current version without it?

**Comment by AngieKadnar** [comment]
@maraf What exactly do you mean?

With this feature, it is now possible to open code.workspace files as an administrator in VSCode.  Previously, this was only possible by specifying the path to VSCode, so your go to IDE is VSCode. However, if you want to open sln files with an other IDE (Visual Studio or Rider) and only the code.workspace files with VSCode, then this is now possible as an administrator.

**Comment by maraf** [comment]
@AngieKadnar I see, thanks. The feature requires to have VSCode on `$path`. If you don't, it doesn't work the same way it doesn't work for VS. On the other hand, it's true that VSCode's default instalation adds `code` alias. I'm just trying to get consistent behavior

File: src/GitExtensions.SolutionRunner/UI/SolutionItemMenuItem.cs [hunk 1]
```diff
@@ -35,11 +38,32 @@ protected override void OnClick(EventArgs e)
...
+                process.StartInfo.FileName = VsCodeExe;
+
+            process.StartInfo.Verb = RunAsAdminConfiguration;
+
+            try
+            {
+                process.Start();
+            }
+            catch (Win32Exception ex)
+            {
```
Comment (maraf) [inline src/GitExtensions.SolutionRunner/UI/SolutionItemMenuItem.cs:68]: Please don't swallow the error. If user configures "runas" and don't provide path to VS, clicking on the solution file does nothing.

So ideally check executable is not set, it's not a VSCode workspace and runas is set and show some messagebox with explanation

**Comment by maraf** [comment]
I have found a "nice" solution. With a bit of registry magic, we can find a system prefered executable for file extension

```c#
string? FindRunTemplate(string fileExtension)
{
    // Find registry key for file extension
    var extensionKey = Registry.ClassesRoot.OpenSubKey(fileExtension);
    var programId = extensionKey.GetValue(String.Empty);
    if (programId == null)
    {
        // If the registry doesn't have a single associated program id,
        // try to search in subkey 'OpenWithProgids'

        extensionKey = extensionKey.OpenSubKey("OpenWithProgids");
        programId = extensionKey.GetValueNames().FirstOrDefault();
        if (programId == null)
        {
            // Out of options
            extensionKey.Close();
            return null;
        }
    }

    // Find run template in form of "C:\Program Files (x86)\Common Files\Microsoft Shared\MSEnv\VSLauncher.exe" "%1"
    var programKey = Registry.ClassesRoot.OpenSubKey(programId + @"\shell\open\command");
    var runTemplate = programKey.GetValue(String.Empty);
    extensionKey.Close();
    programKey.Close();
    return runTemplate.ToString();
}
```

It gives results like
```
> FindRunTemplate(".doc")
"C:\Program Files\Microsoft Office\Root\Office16\WINWORD.EXE" /n "%1" /o "%u"
> FindRunTemplate(".code-workspace")
"C:\Users\marekfisera\AppData\Local\Programs\Microsoft VS Code\Code.exe" "%1"
> FindRunTemplate(".sln")
"C:\Program Files (x86)\Common Files\Microsoft Shared\MSEnv\VSLauncher.exe" "%1"
```

Now we can just replace %1 with selected solution/workspace file

File: src/GitExtensions.SolutionRunner/UI/SolutionItemMenuItem.cs (hunk 1 above)
Comment (AngieKadnar) [reply src/GitExtensions.SolutionRunner/UI/SolutionItemMenuItem.cs:68]: Thanks for the input, it was a quick fix and forgot to add here code :). I add an message box :)

**Comment by AngieKadnar** [comment]
> I have found a "nice" solution. With a bit of registry magic, we can find a system prefered executable for file extension
> 
> ```cs
> string? FindRunTemplate(string fileExtension)
> {
>     // Find registry key for file extension
>     var extensionKey = Registry.ClassesRoot.OpenSubKey(fileExtension);
>     var programId = extensionKey.GetValue(String.Empty);
>     if (programId == null)
>     {
>         // If the registry doesn't have a single associated program id,
>         // try to search in subkey 'OpenWithProgids'
> 
>         extensionKey = extensionKey.OpenSubKey("OpenWithProgids");
>         programId = extensionKey.GetValueNames().FirstOrDefault();
>         if (programId == null)
>         {
>             // Out of options
>             extensionKey.Close();
>             return null;
>         }
>     }
> 
>     // Find run template in form of "C:\Program Files (x86)\Common Files\Microsoft Shared\MSEnv\VSLauncher.exe" "%1"
>     var programKey = Registry.ClassesRoot.OpenSubKey(programId + @"\shell\open\command");
>     var runTemplate = programKey.GetValue(String.Empty);
>     extensionKey.Close();
>     programKey.Close();
>     return runTemplate.ToString();
> }
> ```
> 
> It gives results like
> 
> ```
> > FindRunTemplate(".doc")
> "C:\Program Files\Microsoft Office\Root\Office16\WINWORD.EXE" /n "%1" /o "%u"
> > FindRunTemplate(".code-workspace")
> "C:\Users\marekfisera\AppData\Local\Programs\Microsoft VS Code\Code.exe" "%1"
> > FindRunTemplate(".sln")
> "C:\Program Files (x86)\Common Files\Microsoft Shared\MSEnv\VSLauncher.exe" "%1"
> ```
> 
> Now we can just replace %1 with selected solution/workspace file

Very nice :). I try this solution. Thanks for the input :)

File: src/GitExtensions.SolutionRunner/UI/SolutionItemMenuItem.cs (hunk 1 above)
Comment (AngieKadnar) [reply src/GitExtensions.SolutionRunner/UI/SolutionItemMenuItem.cs:68]: With the latest fix, the try-catch block is no longer necessary, so I have removed it :)

**Comment by AngieKadnar** [comment]
@maraf I implemented the given solution. It works very smooth now :)

File: src/GitExtensions.SolutionRunner/UI/SolutionItemMenuItem.cs [hunk 2]
```diff
@@ -41,5 +46,49 @@ protected override void OnClick(EventArgs e)
...
+        {
+            if(programId == null)
+                return null;
+
+            var programKey = Registry.ClassesRoot.OpenSubKey(programId + @"\shell\open\command");
+            var runTemplate = GetAssociatedProgramId(programKey);
+
+            programKey?.Close();
+
+            return runTemplate?.ToString()?.Replace(" \"%1\"", "");
```
Comment (maraf) [inline src/GitExtensions.SolutionRunner/UI/SolutionItemMenuItem.cs]: The `%1` is a placeholder for file to open. Both VS and VSCode (at the moment) has only a single argument when opening a file, so this works for them. But we should replace `%1` with actual file to open to be correct

**Review by maraf — COMMENTED** [review]
Good job 👍
I'm sorry for late reply, busy times

File: src/GitExtensions.SolutionRunner/UI/SolutionItemMenuItem.cs (hunk 2 above)
Comment (maraf) [reply src/GitExtensions.SolutionRunner/UI/SolutionItemMenuItem.cs]: Good example is a `.doc`

```
> FindRunTemplate(".doc")
"C:\Program Files\Microsoft Office\Root\Office16\WINWORD.EXE" /n "%1" /o "%u"
```

Comment (maraf) [reply src/GitExtensions.SolutionRunner/UI/SolutionItemMenuItem.cs]: We will have split the "template" to `app` and `args` parts and feed them separately to the process start info

**Review by maraf — APPROVED** [review]
Looks great 👍 Job well done and thank you for responding to feedback!

**Pull request merged**